src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -1648,6 +1648,19 @@ data class TyWithObligations<out T>(
 fun <T> TyWithObligations<T>.withObligations(addObligations: List<Obligation>) =
     TyWithObligations(value, obligations + addObligations)
 
+private fun Ty.stripReferences(): Pair<Ty, Mutability?> {
+    var isMut = true
+    var ty = this
+    while (ty is TyReference) {
+        isMut = isMut && ty.mutability.isMut
+        ty = ty.referenced
+    }
+    return ty to (if (this is TyReference) Mutability.valueOf(isMut) else null)
+}
+
+private fun Ty.toRef(mut: Mutability?): Ty =
```

Comment:
I don't like method name because its calls make me think that it always converts type to reference type
Maybe rename it to something like `toRefIfNeeded` (really, I don't like this name - it's too hard to read) and move it into `RsPat.extractBindings` because this method is too specific and should be used only in `extractBindings`